pages/data.py

```diff
@@ -62,142 +113,262 @@ def update_store_trajectories(start_date: str, end_date: str, tz: str, excluded_
+# Controls visibility of columns in trips table and updates the label of button based on the number of clicks.
 def update_dropdowns_trips(n_clicks, button_label):
-    if n_clicks % 2 == 0:
-        hidden_col = ["data.duration_seconds", "data.distance_meters","data.distance"]
-        button_label = 'Display columns with raw units'
-    else:
-        hidden_col = ["data.duration", "data.distance_miles", "data.distance_km", "data.distance"]
-        button_label = 'Display columns with humanzied units'
-    #return the list of hidden columns and the updated button label
+    with ect.Timer() as total_timer:
+
+        # Stage 1: Determine hidden columns and button label based on number of clicks
+        with ect.Timer() as stage1_timer:
+            if n_clicks % 2 == 0:
+                hidden_col = ["data.duration_seconds", "data.distance_meters", "data.distance"]
+                button_label = 'Display columns with raw units'
+            else:
+                hidden_col = ["data.duration", "data.distance_miles", "data.distance_km", "data.distance"]
+                button_label = 'Display columns with humanized units'
+        esdsq.store_dashboard_time(
+            "admin/data/update_dropdowns_trips/determine_hidden_columns_and_label",
+            stage1_timer
+        )
+
+    # Store the total time for the entire function
+    esdsq.store_dashboard_time(
+        "admin/data/update_dropdowns_trips/total_time",
+        total_timer
+    )
+
+    # Return the list of hidden columns and the updated button label
     return hidden_col, button_label
 
 
+
 def populate_datatable(df, table_id=''):
-    if not isinstance(df, pd.DataFrame):
-        raise PreventUpdate
-    return dash_table.DataTable(
-        id= table_id,
-        # columns=[{"name": i, "id": i} for i in df.columns],
-        data=df.to_dict('records'),
-        export_format="csv",
-        filter_options={"case": "sensitive"},
-        # filter_action="native",
-        sort_action="native",  # give user capability to sort columns
-        sort_mode="single",  # sort across 'multi' or 'single' columns
-        page_current=0,  # page number that user is on
-        page_size=50,  # number of rows visible per page
-        style_cell={
-            'textAlign': 'left',
-            # 'minWidth': '100px',
-            # 'width': '100px',
-            # 'maxWidth': '100px',
-        },
-        style_table={'overflowX': 'auto'},
-        css=[{"selector":".show-hide", "rule":"display:none"}]
+    with ect.Timer() as total_timer:
+
+        # Stage 1: Check if df is a DataFrame and raise PreventUpdate if not
+        with ect.Timer() as stage1_timer:
+            if not isinstance(df, pd.DataFrame):
+                raise PreventUpdate
+        esdsq.store_dashboard_time(
+            "admin/data/populate_datatable/check_dataframe_type",
+            stage1_timer
+        )
```

Comment:
I am going to merge this now, but I am not sure we need to instrument every single line going forward. This is not doing any data access or any complex computation; it is just an instance check. It should be consistently fast, although it doesn't hurt to check.